guidance/_chat.py

```diff
@@ -0,0 +1,210 @@
+import warnings
+import uuid
+import inspect
+
+class ChatTemplate:
+    """Contains template for all chat and instruct tuned models."""
+
+    def get_role_start(self, role_name, **kwargs):
+        raise NotImplementedError(
+            "You need to use a ChatTemplate subclass that overrides the get_role_start method"
+        )
+
+    def get_role_end(self, role_name=None):
+        raise NotImplementedError(
+            "You need to use a ChatTemplate subclass that overrides the get_role_start method"
+        )
+    
+class ChatTemplateCache: 
+    def __init__(self):
+        self._cache = {}
+
+    def __getitem__(self, key):
+        key_compact = key.replace(" ", "")
```

Comment:
Minor point: will this collapse multiple consecutive spaces?

Reply:
nah it's a good catch, I'm actually not sure we need to be doing this. I didn't want minor differences in jinja formats (which I believe are whitespace agnostic for parts of them) to cause different mappings in the cache, but maybe there are places where an extra space is actually a meaningful difference?

Reply:
That was a question I asked before I caught sight of the actual keys you were using....